### ISSUE TITLE
Fast Feed-Forward Networks do not perform as well as Normal Feed-Forward Networks

### ISSUE THREAD

**Issue by catid** [pbelcak/fastfeedforward#8]
The code in this repo was missing any comparison with a baseline FFN of the same size, so I added it here: https://github.com/catid/fastfeedforward/blob/main/notebook/example.ipynb

Results:

```
Normal FFN
Number of model parameters: 101770
Evaluation speed: 195.72it/s
Best average accuracy achieved: 0.9785403481012658
```

```
Proposed FFF
Number of model parameters: 107336
Evaluation speed: 111.61it/s
Best average accuracy achieved: 0.9294897151898734
```

Conclusions:

Proposed FFF networks do not perform as well as baseline FFNs in any metric.

I have provided more benchmarks with FFF layers for CIFAR-10 task and demonstrated they do not perform as well as small FFN networks that are faster at inference, negating the stated performance advantages of FFF layers.  More details here: https://github.com/pbelcak/UltraFastBERT/issues/12

**Comment by pbelcak** [comment]
Hi Chris,

This is a third issue in this spirit that you've opened, and for the third time, I'm afraid that the only conclusion is that you're missing the point. 

The behaviour of FFFs as a FF replacement, including a detailed description of their peculiarities and pitfalls of their usage (e.g. overfragmentation) is given in [the original paper](https://arxiv.org/pdf/2308.14711.pdf). Please note **carefully** the claims about their predictive performance made in the paper as well as the detailed evaluation in several settings.

You're more than welcome to produce a reproducibility report on the results reported within, but note that this requires a certain degree of scientific discipline, namely fully adhering to the method outlined in the paper, preserving all hyperparameter settings, and running repeated trials.

To save you the effort of reading through the paper: nowhere has the claim been made that FFFs achieve the same predictive performance under exactly the same setting. 

I'm sure you will be pleased to find out that at the loss of a few % of GLUE performance, FFFs can indeed take the place of FFs in language models ([paper](https://arxiv.org/abs/2311.10770)).

**Comment by catid** [comment]
What I've demonstrated already is the FFF networks perform as well as much smaller standard FF networks.  Therefore, you do not gain anything by using FFF.  For example, I expect you can achieve the same GLUE performance as your FFF modification of BERT using a much smaller FF network with many times fewer weights.  Usually about 1-2x the number of actual evaluations of FFF.

Using these smaller FFNs yields faster inference, faster training, and better accuracy scores in my testing.

Based on your paper you have not performed a baseline against small FF networks to check if your results are good or not.

**Comment by catid** [comment]
I'm not saying that FFF is a bad idea in general.  Perhaps with proper regularization and training modifications, it could work.  But so far I've yet to see this architecture produce useful results in any setting.
